lib/rex/ui/text/shell/history_manager.rb

```diff
@@ -0,0 +1,74 @@
+# -*- coding: binary -*-
+
+module Rex
+module Ui
+module Text
+module Shell
+
+class HistoryManager
+ 
+
+  @@contexts = [{:history_file => Msf::Config.history_file, :name => :msfconsole}]
+  
+  def self.inspect
+    "#<HistoryManager stack size: #{@@contexts.length}>"
+  end
+
+  def self.context_stack
+    @@contexts
+  end
+
+  def self.push_context(history_file: nil, name: nil)
+    return if @@contexts[-1][:name] == name
```

Comment:
This is a little concerning because it seems to imply that the History Manager is anticipating that it could be called multiple times to push a context onto the stack which it may not be necessary. It seems like that would imply that this is a bug in the callers. If a duplicate context is being pushed onto the stack that's the same as what is currently at the top, then it seems like that duplicate shouldn't be being pushed. Likewise the caller shouldn't be inspecting the stack but rather strive to be the one point at which it's necessary to set the context.

Reply:
I've updated this. Thanks for the solution! :+1:

Reply:
I used the `interact` method for pushing the context as you suggested but was not able to use the `detach` method while poping because it was not being called after the session was going to background or exiting, not sure why.